### PR TITLE
EMG: Adding intra-sessions classification with tangent space projection

### PR DESCRIPTION
**Pull request by MariusGuerard**
This PR add a new section in notebook 07, comparing two different algorithms (Logistic Regression and DNN) applied into the vectorized SPD (Euclidean) or into the tangent space projection of the SPD (Riemannian),

### CODE REVIEW

**Comment by ninamiolane** [comment]
Travis spotted an error while running the notebook:
```
KeyError: "Passing list-likes to .loc or [] with any missing labels is no longer supported. The following labels were missing: Int64Index([ 31470,  31570,  31670,  31770,  31870,\n            ...\n            553717, 553817, 553917, 554017, 554117],\n           dtype='int64', length=732). See https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#deprecate-loc-reindex-listlike"
KeyError: "Passing list-likes to .loc or [] with any missing labels is no longer supported. The following labels were missing: Int64Index([ 31470,  31570,  31670,  31770,  31870,\n            ...\n            553717, 553817, 553917, 554017, 554117],\n           dtype='int64', length=732). See https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#deprecate-loc-reindex-listlike"
```
See: https://travis-ci.org/github/geomstats/geomstats/jobs/759593812

File: notebooks/07_emg_hand_gesture_classification.ipynb [hunk 1]
```diff
@@ -117,7 +117,8 @@
    "metadata": {},
```
Comment (ninamiolane) [inline notebooks/07_emg_hand_gesture_classification.ipynb:117]: <p>how we *can accurately* discriminate <em>between</em> these 4 hand *positions.*</p><p> </p><p>The baseline accuracy is defined as the accuracy we get by randomly guessing the signs. In our case, the baseline accuracy is 25%.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/geomstats/geomstats/pull/912/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='32'/>

Comment (ninamiolane) [inline notebooks/07_emg_hand_gesture_classification.ipynb:117]: <p>NIT: Would it be possible to name the pipelines a bit more explicitly? "lr" can be confused with learning rate, ai with artificial intelligence, etc, for users who would read fast through the notebook ... maybe names like:</p><p>logistic_regr__eucl?</p><p>tangent_logistic_regr_affine_invariant?</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/geomstats/geomstats/pull/912/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='42'/>

File: geomstats/datasets/prepare_emg_data.py [hunk 2]
```diff
@@ -17,6 +17,8 @@
 class TimeSeriesCovariance:
     """Class for generating a list of covariance matrices from time series.
 
+    Prepare a TimeSeriesCovariance Object from time series in dictionary.
+
     Parameters
     ----------
```
Comment (ninamiolane) [inline geomstats/datasets/prepare_emg_data.py:23]: Sorry I didn't see this earlier, but could we replace `time_vec` by `time`? `time_vec` seems less natural there, and you also use `time` in the emg.csv data. It can seem like a small detail, but using `time` everywhere will be more intuitive and consistent, thus improving the user experience. Thanks!

Comment (ninamiolane) [reply geomstats/datasets/prepare_emg_data.py:23]: Second NIT on code design/organization: could we rename `n_elecs` into `n_timeseries`? 

Since your code works for any multi-timeseries, we will later add it to the module that transforms data into data on manifolds:
- e.g. put graphs on hyperbolic spaces,
- e.g. put multi-timeseries on spd manifolds, through their covariance matrices.
Thus, if we start writing general code, that's awesome and will save us time down the road!

File: notebooks/07_emg_hand_gesture_classification.ipynb (hunk 1 above)
Comment (ninamiolane) [inline notebooks/07_emg_hand_gesture_classification.ipynb:117]: <p>Naming: the standard is to use the English order:</p><p>pipeline's name -&gt; pipeline_name</p><p>exp's result -&gt; exp_result</p><p>etc</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/geomstats/geomstats/pull/912/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='38'/>

Comment (ninamiolane) [inline notebooks/07_emg_hand_gesture_classification.ipynb:117]: <p>Code design: Just 2 complementary proposals, I let you decide if you feel this is important enough to dedicate your time to it.</p><p> </p><p>Proposal 1:</p><p>Would it make more sense to think about an ExpResults object as being linked to one single experiment, and called Experiment? Thus, with 4 experiments, you would instantiate 4 objects. This way, instead of having dictionaries of lists within a single big object, you have one list of multiple objects with attributes.</p><p> </p><p>The attributes of the object would be:</p><ul><li>self.name (name of the experiment), eg mg_*</li><li>self.ids or directly self.time_series?</li><li>self.pipeline?</li><li>self.session_type: either intra session or inter session, or maybe "self.train_session_id", "self.test_session_id"</li><li>self.subject_type: either intra subject or inter subject, or maybe "self.train_subject_id", "self.test_subject_id"</li><li>self.fit_time</li><li>self.score_time</li><li>self.test_score</li><li>self.train_score</li></ul><p>we could also add:</p><ul><li>self.statistical_method</li><li>self.geometric_method</li></ul><p>thinking forward about the possibility of comparing the impact of statistics versus geometry,</p><p>and individual methods:</p><ul><li>self.run(): would run cross_validate</li><li>self.print_result:</li><li>self.plot_result: would return an ax, so that plots of multiple experiments can be combined.</li></ul><p> </p><p>The method plot_results, would become a function, that either takes a list of ax as inputs, or a list of results.</p><p> </p><p>Proposal 2: Use MLFLOW</p><p> </p><p>Mlflow actually does exactly this. It is made to record experiments, their score, their attributes...and it provides multiple comparison tools right away. I am not sure that we want to add a dependency on mlflow to geomstats at this point, but you could install it in the notebook with:</p><p><code>!pip install mlflow</code> and use it. See:</p><p>https://www.mlflow.org/docs/latest/concepts.html</p><p>https://github.com/mlflow/mlflow/blob/master/examples/sklearn_logistic_regression/train.py</p><p> </p><p>If you want to log the results using mlflow, we could add a method:</p><ul><li>self.mlflow_log()</li></ul><p>that would log the different scores, metrics, and plots to mlflow.</p><p> </p><p>If you choose not to go with the object structure above nor mlflow, could you document what is expected as keys and values in the dictionary data structures, eg adding the Parameters section in the docstring for example. Otherwise it is hard to follow which keys are expected in which dictionary etc.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/geomstats/geomstats/pull/912/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='39'/>

Comment (ninamiolane) [inline notebooks/07_emg_hand_gesture_classification.ipynb:117]: <p>With the object-oriented proposal above, this code would be:</p><p> </p><pre class="ql-syntax" spellcheck="false">pipeline = ...
logreg_eucl_exp = Experiment(pipeline)
logreg_eucl_exp.run(cov_data)
logreg_eucl_exp.print_result()
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/geomstats/geomstats/pull/912/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='42'/>

Comment (ninamiolane) [inline notebooks/07_emg_hand_gesture_classification.ipynb:117]: <p>This is actually very interesting! We should document it as clearly as possible.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/geomstats/geomstats/pull/912/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='53'/>

Comment (ninamiolane) [inline notebooks/07_emg_hand_gesture_classification.ipynb:117]: <p>Very nice!</p><p> </p><p>With the object oriented proposal, this would be: <code>plot_results(exp_list, "test_score")</code></p><p> </p><p>Maybe we can remove <code>stats_list</code> as parameter for now, since you don't use it + I don't understand which are the "two variables" that you are referring to in the docstring? It would also leave the question of statistical significance to the next PR.</p><p> </p><p>If we go the route of comparing the effects of statistics versus geometry, we could start organizing the experiments in:</p><ul><li>2D bar plots, where each line represents a geometric method, and each column a statistical method</li><li>Table, where each line represents a geometric method, and each column a statistical method and we report mean, std dev in each cell.</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/geomstats/geomstats/pull/912/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='60'/>

**Comment by ninamiolane** [comment]
This is great 🎉  A really nice situation of how to use geomstats to investigate if geometry and/or statistics is the main deciding factor when one seeks to classify real world data.

Main comments:
- Travis needs to pass
- Time-series cross-validation: Does your timesplit respect the [timeseries structure](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.TimeSeriesSplit.html)?
> See the time-series aware cross-validation scheme from: https://scikit-learn.org/stable/modules/cross_validation.html
"While i.i.d. data is a common assumption in machine learning theory, it rarely holds in practice. If one knows that the samples have been generated using a time-dependent process, it is safer to use a time-series aware cross-validation scheme. Similarly, if we know that the generative process has a group structure (samples collected from different subjects, experiments, measurement devices), it is safer to use group-wise cross-validation."
- Code design related to experiments, and mlflow (see in notebook): could be useful to be organized if we want to go towards publising a review on geometry vs statistics for emg data.

File: notebooks/07_emg_hand_gesture_classification.ipynb (hunk 1 above)
Comment (MariusGuerard) [reply notebooks/07_emg_hand_gesture_classification.ipynb:117]: <p>Sounds good, I think I will do that in an other PR though (maybe something like "Impact of the relation between hyper-parameters and geometry on the accuracy.")</p><p> </p>
